File: packages/research/anomaly/graph_laplacian.py
"""
Graph Laplacian Anomaly Detection.
Based on US9805002B2 (IBM, expired October 31, 2021).

This module implements semi-supervised anomaly detection using
graph Laplacian regularization and latent variable models.

Key innovations from patent:
- Similarity matrix encodes label constraints (normal-normal positive,
  normal-anomalous negative)
- Latent variable model for dimensionality reduction
- Gradient optimization for joint W and Z learning
- Anomaly score via reconstruction error in latent space
"""

import numpy as np
from numpy.typing import NDArray
from typing import Tuple, Optional, List, Dict
from dataclasses import dataclass
from scipy.linalg import eigh
from scipy.sparse import csr_matrix
from scipy.sparse.linalg import eigsh

from ..core.types import CausalGraph, AnomalyScore


@dataclass
class GraphAnomalyConfig:
    """Configuration for graph Laplacian anomaly detection."""
    latent_dim: int = 10  # D' in the patent
    lambda_reg: float = 1.0  # Graph Laplacian regularization strength
    learning_rate: float = 0.01
    max_iter: int = 100
    tol: float = 1e-6
    normal_similarity: float = 1.0  # 'a' in patent
    anomaly_similarity: float = -1.0  # 'b' in patent
    unlabeled_similarity: float = 0.0  # 'c' in patent


def construct_similarity_matrix(
    n_samples: int,
    labeled_normal: List[int],
    labeled_anomaly: List[int],
    config: GraphAnomalyConfig = GraphAnomalyConfig()
) -> NDArray[np.float64]:
    """
    Construct similarity matrix encoding label constraints.

    From US9805002B2:
    - Normal-normal pairs: positive similarity 'a'
    - Normal-anomalous pairs: non-positive similarity 'b'
    - Constraint: b ≤ c ≤ a

    Args:
        n_samples: Total number of samples
        labeled_normal: Indices of samples labeled as normal
        labeled_anomaly: Indices of samples labeled as anomalous
        config: Configuration with similarity values

    Returns:
        Similarity matrix of shape (n_samples, n_samples)
    """
    R = np.full((n_samples, n_samples), config.unlabeled_similarity)

    # Normal-normal pairs
    for i in labeled_normal:
        for j in labeled_normal:
            if i != j:
                R[i, j] = config.normal_similarity

    # Anomaly-anomaly pairs
    for i in labeled_anomaly:
        for j in labeled_anomaly:
            if i != j:
                R[i, j] = config.normal_similarity  # Anomalies similar to each other

    # Normal-anomaly pairs
    for i in labeled_normal:
        for j in labeled_anomaly:
            R[i, j] = config.anomaly_similarity
            R[j, i] = config.anomaly_similarity

    return R


def graph_laplacian(R: NDArray[np.float64]) -> NDArray[np.float64]:
    """
    Compute graph Laplacian from similarity matrix.

    L = D_R - R

    where D_R is diagonal degree matrix.

    Args:
        R: Similarity matrix

    Returns:
        Graph Laplacian matrix
    """
    D = np.diag(R.sum(axis=1))
    return D - R


class GraphLatentAnomalyDetector:
    """
    Latent variable model with graph Laplacian regularization.

    From US9805002B2:
    X_{n,d} = Z_n · W_d + s · ε

    where Z is latent variable, W is sensor-specific coefficients,
    s varies by label status.

    NOVEL INSIGHT #3 from Great Attractor framework:
    Phase transitions are moments when trajectories cross attractor
    basin boundaries. The latent space Z represents the "possibility
    space" - anomalies are points near basin boundaries where the
    gradient field is unstable.
    """

    def __init__(self, config: GraphAnomalyConfig = GraphAnomalyConfig()):
        self.config = config
        self.W: Optional[NDArray] = None  # Sensor coefficients
        self.Z: Optional[NDArray] = None  # Latent variables

    def fit(
        self,
        X: NDArray[np.float64],
        labeled_normal: List[int] = None,
        labeled_anomaly: List[int] = None,
        similarity_matrix: Optional[NDArray[np.float64]] = None
    ) -> 'GraphLatentAnomalyDetector':
        """
        Fit the latent variable model.

        From US9805002B2, gradient updates:
        W := W - α[{S ⊙ (X - ZW^T)}^T Z + N(WW^T)^{-1}W]
        Z := Z - α[{S ⊙ (X - ZW^T)}W + λLZ]

        Args:
            X: Data matrix of shape (N, D)
            labeled_normal: Indices of normal samples
            labeled_anomaly: Indices of anomalous samples
            similarity_matrix: Pre-computed similarity matrix (optional)

        Returns:
            Self
        """
        N, D = X.shape
        D_prime = self.config.latent_dim

        # Initialize
        self.W = np.random.randn(D, D_prime) * 0.1
        self.Z = np.random.randn(N, D_prime) * 0.1

        # Construct similarity and Laplacian
        if similarity_matrix is None:
            labeled_normal = labeled_normal or []
            labeled_anomaly = labeled_anomaly or []
            R = construct_similarity_matrix(N, labeled_normal, labeled_anomaly, self.config)
        else:
            R = similarity_matrix

        L = graph_laplacian(R)

        # Scale matrix (1 for all in unsupervised case)
        S = np.ones_like(X)

        # Gradient descent
        for iteration in range(self.config.max_iter):
            # Reconstruction error
            residual = X - self.Z @ self.W.T
            scaled_residual = S * residual

            # W update
            grad_W = -scaled_residual.T @ self.Z
            # Regularization term for numerical stability
            reg_W = self.config.lambda_reg * self.W
            self.W -= self.config.learning_rate * (grad_W + reg_W)

            # Z update with Laplacian regularization
            grad_Z = -scaled_residual @ self.W + self.config.lambda_reg * L @ self.Z
            self.Z -= self.config.learning_rate * grad_Z

            # Check convergence
            loss = np.sum(scaled_residual ** 2) + self.config.lambda_reg * np.trace(self.Z.T @ L @ self.Z)
            if iteration > 0 and abs(prev_loss - loss) < self.config.tol:
                break
            prev_loss = loss

        return self

    def score(self, X: NDArray[np.float64]) -> NDArray[np.float64]:
        """
        Compute anomaly scores via reconstruction error.

        From US9805002B2:
        s_n = (I - W(W^TW)^{-1}W^T) · X_n

        This is the component of X orthogonal to the learned subspace.

        Args:
            X: Data matrix of shape (N, D) or (D,) for single sample

        Returns:
            Anomaly scores of shape (N,)
        """
        if self.W is None:
            raise ValueError("Model not fitted. Call fit() first.")

        if X.ndim == 1:
            X = X.reshape(1, -1)

        # Projection matrix onto orthogonal complement of W's column space
        WtW_inv = np.linalg.pinv(self.W.T @ self.W)
        P_orth = np.eye(X.shape[1]) - self.W @ WtW_inv @ self.W.T

        # Orthogonal component magnitude
        X_orth = X @ P_orth
        scores = np.linalg.norm(X_orth, axis=1)

        return scores

    def get_basin_boundary_proximity(
        self,
        X: NDArray[np.float64]
    ) -> NDArray[np.float64]:
        """
        NOVEL INSIGHT #3: Proximity to attractor basin boundaries.

        Points near basin boundaries have high gradient magnitude in
        the latent space - small perturbations lead to large changes
        in which attractor they fall toward.

        Computed as local gradient magnitude of anomaly score.

        Args:
            X: Data matrix

        Returns:
            Basin boundary proximity scores
        """
        if self.Z is None:
            raise ValueError("Model not fitted.")

        # Compute gradient of score with respect to latent position
        scores = self.score(X)

        # Numerical gradient estimation via finite differences
        eps = 1e-4
        gradients = np.zeros_like(self.Z)

        for d in range(self.Z.shape[1]):
            Z_plus = self.Z.copy()
            Z_plus[:, d] += eps
            X_plus = Z_plus @ self.W.T

            Z_minus = self.Z.copy()
            Z_minus[:, d] -= eps
            X_minus = Z_minus @ self.W.T

            scores_plus = self.score(X_plus)
            scores_minus = self.score(X_minus)

            gradients[:, d] = (scores_plus - scores_minus) / (2 * eps)

        # Gradient magnitude
        return np.linalg.norm(gradients, axis=1)


def spectral_anomaly_detection(
    causal_graph: CausalGraph,
    features: NDArray[np.float64],
    n_components: int = 5
) -> Tuple[NDArray[np.float64], NDArray[np.float64]]:
    """
    Spectral clustering-based anomaly detection using causal graph.

    Uses the Ng-Jordan-Weiss algorithm from US7103225B2 (NEC, expired 2023).

    1. Compute normalized Laplacian from causal graph
    2. Find first k eigenvectors
    3. Cluster in spectral embedding space
    4. Anomalies are points far from cluster centers

    NOVEL INSIGHT #4: Coherence Amplification
    When agents synchronize intentions, they cluster tightly in spectral
    space. Dispersion indicates competing attractors; tight clusters
    indicate amplified coherent intention.

    Args:
        causal_graph: CausalGraph with transfer entropy weights
        features: Additional node features, shape (N, D)
        n_components: Number of spectral components

    Returns:
        Tuple of (spectral embedding, anomaly scores)
    """
    L_sym = causal_graph.get_normalized_laplacian()

    # Eigen decomposition (smallest eigenvalues for Laplacian)
    if L_sym.shape[0] > 100:
        # Use sparse solver for large graphs
        eigenvalues, eigenvectors = eigsh(csr_matrix(L_sym), k=n_components, which='SM')
    else:
        eigenvalues, eigenvectors = eigh(L_sym)
        eigenvectors = eigenvectors[:, :n_components]

    # Normalize rows to unit length
    row_norms = np.linalg.norm(eigenvectors, axis=1, keepdims=True)
    row_norms = np.where(row_norms > 0, row_norms, 1.0)
    embedding = eigenvectors / row_norms

    # Anomaly score: distance from centroid
    centroid = embedding.mean(axis=0)
    scores = np.linalg.norm(embedding - centroid, axis=1)

    return embedding, scores


def detect_coherence_amplification(
    embeddings_over_time: List[NDArray[np.float64]]
) -> NDArray[np.float64]:
    """
    NOVEL INSIGHT #4: Detect coherence amplification over time.

    When distributed agents synchronize, the spectral embedding becomes
    more concentrated. This indicates an amplified attractor well forming.

    Args:
        embeddings_over_time: List of spectral embeddings at each timestep

    Returns:
        Coherence amplification signal (negative = dispersion, positive = concentration)
    """
    dispersions = []

    for embedding in embeddings_over_time:
        # Measure dispersion as trace of covariance
        centered = embedding - embedding.mean(axis=0)
        cov = centered.T @ centered / len(centered)
        dispersion = np.trace(cov)
        dispersions.append(dispersion)

    dispersions = np.array(dispersions)

    # Coherence = negative rate of change of dispersion
    # Increasing coherence = decreasing dispersion
    coherence = -np.gradient(dispersions)

    return coherence

File: packages/research/rust/src/graph_laplacian.rs
//! Graph Laplacian Anomaly Detection
//!
//! Based on US9805002B2 (IBM, expired October 31, 2021).
//!
//! Implements semi-supervised anomaly detection using:
//! - Graph Laplacian regularization
//! - Latent variable models
//! - Spectral clustering for basin boundary detection

use ndarray::{Array1, Array2, Axis};
use serde::{Deserialize, Serialize};

/// Configuration for graph Laplacian anomaly detection
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphAnomalyConfig {
    /// Latent dimension (D' in the patent)
    pub latent_dim: usize,
    /// Graph Laplacian regularization strength
    pub lambda_reg: f64,
    /// Learning rate for gradient descent
    pub learning_rate: f64,
    /// Maximum iterations
    pub max_iter: usize,
    /// Convergence tolerance
    pub tol: f64,
    /// Similarity for normal-normal pairs
    pub normal_similarity: f64,
    /// Similarity for normal-anomaly pairs
    pub anomaly_similarity: f64,
    /// Similarity for unlabeled pairs
    pub unlabeled_similarity: f64,
}

impl Default for GraphAnomalyConfig {
    fn default() -> Self {
        Self {
            latent_dim: 10,
            lambda_reg: 1.0,
            learning_rate: 0.01,
            max_iter: 100,
            tol: 1e-6,
            normal_similarity: 1.0,
            anomaly_similarity: -1.0,
            unlabeled_similarity: 0.0,
        }
    }
}

/// Construct similarity matrix encoding label constraints
///
/// From US9805002B2:
/// - Normal-normal pairs: positive similarity 'a'
/// - Normal-anomalous pairs: non-positive similarity 'b'
pub fn construct_similarity_matrix(
    n_samples: usize,
    labeled_normal: &[usize],
    labeled_anomaly: &[usize],
    config: &GraphAnomalyConfig,
) -> Array2<f64> {
    let mut r = Array2::from_elem((n_samples, n_samples), config.unlabeled_similarity);

    // Normal-normal pairs
    for &i in labeled_normal {
        for &j in labeled_normal {
            if i != j && i < n_samples && j < n_samples {
                r[[i, j]] = config.normal_similarity;
            }
        }
    }

    // Anomaly-anomaly pairs
    for &i in labeled_anomaly {
        for &j in labeled_anomaly {
            if i != j && i < n_samples && j < n_samples {
                r[[i, j]] = config.normal_similarity;
            }
        }
    }

    // Normal-anomaly pairs
    for &i in labeled_normal {
        for &j in labeled_anomaly {
            if i < n_samples && j < n_samples {
                r[[i, j]] = config.anomaly_similarity;
                r[[j, i]] = config.anomaly_similarity;
            }
        }
    }

    r
}

/// Compute graph Laplacian from similarity matrix
///
/// L = D_R - R where D_R is the diagonal degree matrix
pub fn graph_laplacian(r: &Array2<f64>) -> Array2<f64> {
    let n = r.nrows();
    let mut laplacian = -r.clone();

    // Add degree matrix on diagonal
    for i in 0..n {
        let degree: f64 = r.row(i).sum();
        laplacian[[i, i]] = degree;
    }

    laplacian
}

/// Normalized graph Laplacian: L_sym = D^(-1/2) L D^(-1/2)
pub fn normalized_laplacian(r: &Array2<f64>) -> Array2<f64> {
    let n = r.nrows();
    let l = graph_laplacian(r);

    // Compute degree vector
    let degrees: Array1<f64> = r.sum_axis(Axis(1));

    // D^(-1/2)
    let d_inv_sqrt: Array1<f64> = degrees.mapv(|d| {
        if d > 1e-10 {
            1.0 / d.sqrt()
        } else {
            0.0
        }
    });

    // L_sym = D^(-1/2) L D^(-1/2)
    let mut l_sym = Array2::zeros((n, n));
    for i in 0..n {
        for j in 0..n {
            l_sym[[i, j]] = d_inv_sqrt[i] * l[[i, j]] * d_inv_sqrt[j];
        }
    }

    l_sym
}

/// Result of anomaly detection
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnomalyResult {
    /// Anomaly scores for each sample
    pub scores: Vec<f64>,
    /// Latent embedding (flattened)
    pub embedding: Vec<f64>,
    /// Latent dimensions
    pub latent_dims: usize,
    /// Basin boundary proximity scores
    pub basin_proximity: Vec<f64>,
}

/// Latent variable model with graph Laplacian regularization
pub struct GraphLatentDetector {
    config: GraphAnomalyConfig,
    w: Option<Array2<f64>>, // Sensor coefficients
    z: Option<Array2<f64>>, // Latent variables
}

impl GraphLatentDetector {
    pub fn new(config: GraphAnomalyConfig) -> Self {
        Self {
            config,
            w: None,
            z: None,
        }
    }

    /// Fit the latent variable model using gradient descent
    ///
    /// From US9805002B2, gradient updates:
    /// W := W - α[{S ⊙ (X - ZW^T)}^T Z + N(WW^T)^{-1}W]
    /// Z := Z - α[{S ⊙ (X - ZW^T)}W + λLZ]
    pub fn fit(
        &mut self,
        x: &Array2<f64>,
        labeled_normal: &[usize],
        labeled_anomaly: &[usize],
    ) {
        let (n, d) = (x.nrows(), x.ncols());
        let d_prime = self.config.latent_dim;

        // Initialize
        use rand::Rng;
        let mut rng = rand::thread_rng();

        let mut w = Array2::from_shape_fn((d, d_prime), |_| rng.gen::<f64>() * 0.1);
        let mut z = Array2::from_shape_fn((n, d_prime), |_| rng.gen::<f64>() * 0.1);

        // Construct similarity and Laplacian
        let r = construct_similarity_matrix(n, labeled_normal, labeled_anomaly, &self.config);
        let l = graph_laplacian(&r);

        let mut prev_loss = f64::INFINITY;

        for _ in 0..self.config.max_iter {
            // Reconstruction error: X - ZW^T
            let reconstruction = z.dot(&w.t());
            let residual = x - &reconstruction;

            // W update
            let grad_w = -residual.t().dot(&z) + &w * self.config.lambda_reg;
            w = w - &grad_w * self.config.learning_rate;

            // Z update with Laplacian regularization
            let grad_z = -residual.dot(&w) + l.dot(&z) * self.config.lambda_reg;
            z = z - &grad_z * self.config.learning_rate;

            // Check convergence
            let loss = residual.mapv(|x| x * x).sum()
                + self.config.lambda_reg * z.t().dot(&l).dot(&z).diag().sum();

            if (prev_loss - loss).abs() < self.config.tol {
                break;
            }
            prev_loss = loss;
        }

        self.w = Some(w);
        self.z = Some(z);
    }

    /// Compute anomaly scores via reconstruction error
    ///
    /// From US9805002B2:
    /// s_n = (I - W(W^TW)^{-1}W^T) · X_n
    pub fn score(&self, x: &Array2<f64>) -> Vec<f64> {
        let w = match &self.w {
            Some(w) => w,
            None => return vec![0.0; x.nrows()],
        };

        let n = x.nrows();
        let d = x.ncols();

        // Compute W(W^TW)^{-1}W^T using pseudo-inverse
        let wtw = w.t().dot(w);

        // Simple regularized inverse
        let mut wtw_inv = Array2::zeros((wtw.nrows(), wtw.ncols()));
        for i in 0..wtw.nrows() {
            for j in 0..wtw.ncols() {
                if i == j {
                    wtw_inv[[i, j]] = 1.0 / (wtw[[i, j]] + 1e-6);
                }
            }
        }

        // Projection matrix onto orthogonal complement
        let p_w = w.dot(&wtw_inv).dot(&w.t());
        let p_orth = Array2::eye(d) - p_w;

        // Orthogonal component magnitude
        let x_orth = x.dot(&p_orth);

        (0..n)
            .map(|i| x_orth.row(i).mapv(|v| v * v).sum().sqrt())
            .collect()
    }

    /// Get latent embedding
    pub fn embedding(&self) -> Option<Vec<f64>> {
        self.z.as_ref().map(|z| z.clone().into_raw_vec())
    }

    /// Compute basin boundary proximity
    ///
    /// Points near basin boundaries have high gradient magnitude
    pub fn basin_proximity(&self, x: &Array2<f64>) -> Vec<f64> {
        let (z, w) = match (&self.z, &self.w) {
            (Some(z), Some(w)) => (z, w),
            _ => return vec![0.0; x.nrows()],
        };

        let n = z.nrows();
        let d_prime = z.ncols();
        let eps = 1e-4;

        let mut gradients = Array2::zeros((n, d_prime));

        for d in 0..d_prime {
            // Perturb in dimension d
            let mut z_plus = z.clone();
            let mut z_minus = z.clone();

            for i in 0..n {
                z_plus[[i, d]] += eps;
                z_minus[[i, d]] -= eps;
            }

            let x_plus = z_plus.dot(&w.t());
            let x_minus = z_minus.dot(&w.t());

            let scores_plus = self.score(&x_plus);
            let scores_minus = self.score(&x_minus);

            for i in 0..n {
                gradients[[i, d]] = (scores_plus[i] - scores_minus[i]) / (2.0 * eps);
            }
        }

        // Gradient magnitude
        (0..n)
            .map(|i| gradients.row(i).mapv(|v| v * v).sum().sqrt())
            .collect()
    }
}

/// Spectral anomaly detection using eigendecomposition
///
/// Uses the Ng-Jordan-Weiss algorithm from US7103225B2 (NEC, expired 2023)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpectralAnomalyResult {
    /// Spectral embedding
    pub embedding: Vec<f64>,
    /// Embedding dimensions
    pub n_dims: usize,
    /// Number of samples
    pub n_samples: usize,
    /// Anomaly scores
    pub scores: Vec<f64>,
    /// Eigenvalues used
    pub eigenvalues: Vec<f64>,
}

/// Compute spectral embedding and anomaly scores
///
/// 1. Compute normalized Laplacian from similarity matrix
/// 2. Find first k eigenvectors
/// 3. Anomalies are points far from embedding centroid
pub fn spectral_anomaly_detection(
    similarity: &Array2<f64>,
    n_components: usize,
) -> SpectralAnomalyResult {
    let n = similarity.nrows();
    let l_sym = normalized_laplacian(similarity);

    // Power iteration for dominant eigenvectors
    // (Simplified - for production use nalgebra or ndarray-linalg)
    let k = n_components.min(n);
    let mut embedding = Array2::zeros((n, k));
    let mut eigenvalues = Vec::with_capacity(k);

    // Initialize with random vectors
    use rand::Rng;
    let mut rng = rand::thread_rng();

    for comp in 0..k {
        let mut v = Array1::from_shape_fn(n, |_| rng.gen::<f64>() - 0.5);

        // Power iteration for smallest eigenvalue (largest for I - L_sym)
        let shifted = Array2::eye(n) - &l_sym;

        for _ in 0..50 {
            v = shifted.dot(&v);

            // Orthogonalize against previous eigenvectors
            for prev in 0..comp {
                let prev_vec = embedding.column(prev);
                let proj: f64 = v.iter().zip(prev_vec.iter()).map(|(a, b)| a * b).sum();
                v = v - &(prev_vec.to_owned() * proj);
            }

            // Normalize
            let norm: f64 = v.iter().map(|x| x * x).sum::<f64>().sqrt();
            if norm > 1e-10 {
                v = v / norm;
            }
        }

        // Compute eigenvalue (Rayleigh quotient)
        let av = l_sym.dot(&v);
        let eigenvalue: f64 = v.iter().zip(av.iter()).map(|(a, b)| a * b).sum();
        eigenvalues.push(eigenvalue);

        for i in 0..n {
            embedding[[i, comp]] = v[i];
        }
    }

    // Normalize rows to unit length
    for i in 0..n {
        let norm: f64 = embedding.row(i).mapv(|x| x * x).sum().sqrt();
        if norm > 1e-10 {
            for j in 0..k {
                embedding[[i, j]] /= norm;
            }
        }
    }

    // Anomaly score: distance from centroid
    let centroid: Array1<f64> = embedding.mean_axis(Axis(0)).unwrap();
    let scores: Vec<f64> = (0..n)
        .map(|i| {
            let row = embedding.row(i);
            row.iter()
                .zip(centroid.iter())
                .map(|(a, b)| (a - b).powi(2))
                .sum::<f64>()
                .sqrt()
        })
        .collect();

    SpectralAnomalyResult {
        embedding: embedding.into_raw_vec(),
        n_dims: k,
        n_samples: n,
        scores,
        eigenvalues,
    }
}

/// Detect coherence amplification over time
///
/// When distributed agents synchronize, the spectral embedding becomes
/// more concentrated. This indicates an amplified attractor well forming.
pub fn detect_coherence_amplification(dispersions_over_time: &[f64]) -> Vec<f64> {
    if dispersions_over_time.len() < 2 {
        return vec![0.0; dispersions_over_time.len()];
    }

    // Coherence = negative gradient of dispersion
    // Increasing coherence = decreasing dispersion
    let mut coherence = Vec::with_capacity(dispersions_over_time.len());

    // First derivative using central differences
    coherence.push(dispersions_over_time[0] - dispersions_over_time[1]);

    for i in 1..dispersions_over_time.len() - 1 {
        let grad = (dispersions_over_time[i - 1] - dispersions_over_time[i + 1]) / 2.0;
        coherence.push(grad);
    }

    let n = dispersions_over_time.len();
    coherence.push(dispersions_over_time[n - 2] - dispersions_over_time[n - 1]);

    coherence
}

/// Compute embedding dispersion (trace of covariance)
pub fn embedding_dispersion(embedding: &[f64], n_samples: usize, n_dims: usize) -> f64 {
    if n_samples == 0 || n_dims == 0 {
        return 0.0;
    }

    // Compute mean per dimension
    let mut mean = vec![0.0; n_dims];
    for i in 0..n_samples {
        for j in 0..n_dims {
            mean[j] += embedding[i * n_dims + j];
        }
    }
    for j in 0..n_dims {
        mean[j] /= n_samples as f64;
    }

    // Compute trace of covariance (sum of variances)
    let mut trace = 0.0;
    for j in 0..n_dims {
        let mut variance = 0.0;
        for i in 0..n_samples {
            let diff = embedding[i * n_dims + j] - mean[j];
            variance += diff * diff;
        }
        trace += variance / n_samples as f64;
    }

    trace
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_graph_laplacian() {
        let r = Array2::from_shape_vec(
            (3, 3),
            vec![0.0, 1.0, 0.5, 1.0, 0.0, 0.5, 0.5, 0.5, 0.0],
        )
        .unwrap();

        let l = graph_laplacian(&r);

        // Check diagonal is degree
        assert!((l[[0, 0]] - 1.5).abs() < 1e-10);
        assert!((l[[1, 1]] - 1.5).abs() < 1e-10);
        assert!((l[[2, 2]] - 1.0).abs() < 1e-10);

        // Off-diagonal is negative similarity
        assert!((l[[0, 1]] + 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_similarity_matrix() {
        let r = construct_similarity_matrix(
            5,
            &[0, 1],
            &[3, 4],
            &GraphAnomalyConfig::default(),
        );

        // Normal-normal should be positive
        assert!(r[[0, 1]] > 0.0);

        // Normal-anomaly should be negative
        assert!(r[[0, 3]] < 0.0);
        assert!(r[[1, 4]] < 0.0);

        // Unlabeled should be zero
        assert_eq!(r[[0, 2]], 0.0);
    }

    #[test]
    fn test_coherence_amplification() {
        // Decreasing dispersion = increasing coherence
        let dispersions = vec![1.0, 0.8, 0.6, 0.4, 0.2];
        let coherence = detect_coherence_amplification(&dispersions);

        // All should be positive (dispersion decreasing)
        for c in &coherence[1..coherence.len() - 1] {
            assert!(*c > 0.0, "Expected positive coherence, got {}", c);
        }
    }
}
